Support fx_pytree in dynamo

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105574

This PR does two things:
1. Make dynamo trace through fx_pytree (on top of torch.utils._pytree) so that generated graph modules can be retraced. 
2. Fix bug where unflatten not returning dynamo VariableTracker. 

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

Differential Revision: [D47734623](https://our.internmc.facebook.com/intern/diff/D47734623)

tugsbayasgalan committed:
Support fx_pytree in dynamo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support fx_pytree in dynamo"


This PR does two things:
1. Make dynamo trace through fx_pytree (on top of torch.utils._pytree) so that generated graph modules can be retraced. 
2. Fix bug where unflatten not returning dynamo VariableTracker. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

yanboliang commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -244,6 +249,17 @@ def call_method(
+                return ConstantVariable(
+                    args[0].as_python_constant() in self.value, **options
+                ).add_guard(self.source.make_guard(GuardBuilder.ODICT_KEYS))
+
```
L240-L262 are pretty similar logic for ```dict``` and ```OrderedDict```, can you consolidate them?

tugsbayasgalan replied:
Yep!

tugsbayasgalan commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -235,7 +240,7 @@ def call_method(
                 method is collections.OrderedDict.__contains__
                 and len(args) == 1
-                and isinstance(args[0], ConstantVariable)
+                and isinstance(args[0], (ConstantVariable, BuiltinVariable))
```
Yep

tugsbayasgalan replied:
To make sure something like this works:
```py
look_up_dict = {tuple: 4}
a = (5, 6)
type(a) in look_up_dict
```

tugsbayasgalan commented on test/dynamo/test_export.py:
```diff
@@ -3237,6 +3237,89 @@ def forward(self, l_x_, ones_3_true_branch, ones_1_true_branch, ones_true_branch
     return add_2""",
         )
 
+    def test_fx_pytree(self):
```
Sure

tugsbayasgalan committed:
Update on "Support fx_pytree in dynamo"


This PR does two things:
1. Make dynamo trace through fx_pytree (on top of torch.utils._pytree) so that generated graph modules can be retraced. 
2. Fix bug where unflatten not returning dynamo VariableTracker. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support fx_pytree in dynamo"


This PR does two things:
1. Make dynamo trace through fx_pytree (on top of torch.utils._pytree) so that generated graph modules can be retraced. 
2. Fix bug where unflatten not returning dynamo VariableTracker. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47734623).

ydwu4 commented on test/dynamo/test_export.py:
```diff
@@ -3303,6 +3303,102 @@ def forward(self, l_x_, ones_3_true_branch, ones_1_true_branch, ones_true_branch
+            return flat_args_fx[0] + flat_args[0]
+
+        inp_container = (torch.randn(20, 16, 50, 100), torch.randn(20, 16, 50, 100))
+        print(foo(inp_container))
```
nit: remove the print

tugsbayasgalan committed:
Update on "Support fx_pytree in dynamo"


This PR does two things:
1. Make dynamo trace through fx_pytree (on top of torch.utils._pytree) so that generated graph modules can be retraced. 
2. Fix bug where unflatten not returning dynamo VariableTracker. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

Differential Revision: [D47734623](https://our.internmc.facebook.com/intern/diff/D47734623)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47734623).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47734623).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Support fx_pytree in dynamo"


This PR does two things:
1. Make dynamo trace through fx_pytree (on top of torch.utils._pytree) so that generated graph modules can be retraced. 
2. Fix bug where unflatten not returning dynamo VariableTracker. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

Differential Revision: [D47734623](https://our.internmc.facebook.com/intern/diff/D47734623)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge -ic

tugsbayasgalan commented:
@pytorchbot merge -i

tugsbayasgalan commented:
@pytorchbot label "topic: not user facing"

tugsbayasgalan commented:
@pytorchbot merge -i
